pkg/streamingpromql/functions/common.go

```diff
@@ -58,10 +59,53 @@ func Passthrough(seriesData types.InstantVectorSeriesData, _ *pooling.LimitingPo
 //   - rangeSeconds: the duration of the range in seconds.
 //   - floatBuffer: a ring buffer containing float values for the step range.
 //   - histogramBuffer: a ring buffer containing histogram values for the step range.
+//   - emitAnnotation: a callback function to emit an annotation for the current series.
 //
 // Returns:
 //   - hasFloat bool: a boolean indicating if a float value is present.
 //   - f float64: the float value.
 //   - h *histogram.FloatHistogram: nil if no histogram is present.
 //   - err error.
-type RangeVectorStepFunction func(step types.RangeVectorStepData, rangeSeconds float64, floatBuffer *types.FPointRingBuffer, histogramBuffer *types.HPointRingBuffer) (f float64, hasFloat bool, h *histogram.FloatHistogram, err error)
+type RangeVectorStepFunction func(
+	step types.RangeVectorStepData,
+	rangeSeconds float64,
+	floatBuffer *types.FPointRingBuffer,
+	histogramBuffer *types.HPointRingBuffer,
+	emitAnnotation EmitAnnotationFunc,
+) (f float64, hasFloat bool, h *histogram.FloatHistogram, err error)
+
+// EmitAnnotationFunc is a function that emits the annotation created by generator.
+type EmitAnnotationFunc func(generator AnnotationGenerator)
+
+// AnnotationGenerator is a function that returns an annotation for the given metric name and expression position.
+type AnnotationGenerator func(metricName string, expressionPosition posrange.PositionRange) error
+
+// RangeVectorSeriesValidationFunction is a function that is called after a series is completed for a function over a range vector.
+type RangeVectorSeriesValidationFunction func(seriesData types.InstantVectorSeriesData, metricName string, emitAnnotation EmitAnnotationFunc)
+
+// RangeVectorSeriesValidationFunctionFactory is a factory function that returns a RangeVectorSeriesValidationFunction
+type RangeVectorSeriesValidationFunctionFactory func() RangeVectorSeriesValidationFunction
+
+type FunctionOverRangeVector struct {
+	// StepFunc is the function that computes an output sample for a single step.
+	StepFunc RangeVectorStepFunction
+
+	// SeriesValidationFuncFactory is the function that creates a validator for a complete series, emitting any annotations
+	// for that series.
+	//
+	// A new validator instance is created for each instance of this range vector function (ie. a validator is not shared
+	// between queries or between different invocations of the same range vector function in a single query). A single
+	// instance of a range vector function can then implement optimisations such as skipping checks for repeated metric
+	// names.
+	//
+	// SeriesValidationFuncFactory can be nil, in which case no validation is performed.
+	SeriesValidationFuncFactory RangeVectorSeriesValidationFunctionFactory
+
+	// SeriesMetadataFunc is the function that computes the output series for this function based on the given input series.
+	SeriesMetadataFunc SeriesMetadataFunction
```

Comment:
I wonder if we want to similarly create a `FunctionOverInstantVector` for consistency that has `SeriesMetadataFunc` etc.

Reply:
Yep, that's what I was thinking too. But I think that's something for a follow up PR.